components/brave_sync/brave_sync_service_impl.cc

```diff
@@ -488,14 +501,13 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
         record->action == jslib::SyncRecord::Action::A_DELETE &&
           actually_merged;
     }
-  } // for each device
+  }  // for each device
 
   sync_prefs_->SetSyncDevices(*sync_devices);
 
-  if (this_device_deleted)
+  if (this_device_deleted || contains_only_one_device) {
```

Comment:
`ResetSyncInternal` won't send delete record to the sync chain so other devices won't know this device is deleted.

Reply:
this change is inside of BraveSyncServiceImpl::OnResolvedPreferences , where only resolved message comes.

1. if this_device_deleted is true - this means that either our machine or some another already sent DELETE record

2. contains_only_one_device is true - this flag is set when some device had been deleted from chain:
```
      contains_only_one_device = sync_devices->size() < 2 &&
        record->action == jslib::SyncRecord::Action::A_DELETE &&
          actually_merged;
```
so the chain is about to be destroyed, so no need to send anything to sync.
This is the reason why I had removed `OnResetSync();`.

Or what is your concern @darkdh ?

Reply:
I was worried about case 2. 
Considering this,
When we resolved record and find out this device is the only one on chain, and then we reset sync without sending delete record. If users remember the sync code, they can add a new device to sync chain and there will always be a zombie device in sync chain.

Reply:
We might need to implement `DELETE_SYNC_USER` and `DELETED_SYNC_USER` properly when we are about to ditch the sync chain

Reply:
for the case 2
https://github.com/brave/brave-core/pull/1560/commits/6fd03d2f196213597cf9f5e9d10bbc64f04765ba#diff-c04ecef4235e311e0a383fe4fabcb4d4R233
```
void BraveSyncServiceImpl::OnResetSync() {
  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);

  auto sync_devices = sync_prefs_->GetSyncDevices();
  // If there is only one or no devices left, we won't get back resolved sync
  // record back.
  if (sync_devices->size() <= 1) {
    ResetSyncInternal();
  } else {
    // We have to send delete record and wait for library deleted response then
    // we can reset it by ResetInternal()
    const std::string device_id = sync_prefs_->GetThisDeviceId();
    OnDeleteDevice(device_id);
  }
}
```
may then need to either completely remove `ResetSyncInternal();` or to invoke it just in case if (sync_devices->size() == 0)

Reply:
@darkdh 
Updated PR to avoid zombie devices.
Will squash commits after approval for easier re-review.